lr_face/models.py

```diff
@@ -54,5 +55,96 @@ def score_for_pair(self, pair):
         img2 = resize_and_normalize(pair[1], self.model.input_shape[1:3])
         img1_representation = self.model.predict(img1)[0, :]
         img2_representation = self.model.predict(img2)[0, :]
-        score = spatial.distance.cosine(img1_representation, img2_representation)
+        score = spatial.distance.cosine(img1_representation,
+                                        img2_representation)
         return score
+
+
+class FinetuneModel(tf.keras.Model):
+    """
+    A subclass of tf.keras.Model that can be used to finetune the pre-trained
+    embedding models.This new training model takes 3 inputs, namely:
+
+        anchor: A 4D tensor containing a batch of anchor images with shape
+            `(batch_size, height, width, num_channels)`.
+        positive: A 4D tensor containing a batch of images of the same identity
+            as the anchor image with shape `(batch_size, height, width,
+            num_channels)`.
+        positive: A 4D tensor containing a batch of images of a different
+            identity than the anchor image with shape `(batch_size, height,
+            width, num_channels)`.
+
+    It outputs embeddings for each of the images and returns them as a single
+    3D tensor of shape `(batch_size, 3, embedding_size)`, where the second
+    axis represents the anchor, positive and negative images, respectively.
+    The reason for returning the results as a single tensor instead of 3
+    separate outputs is because all 3 are required for computing a single loss.
+    """
+
+    def __init__(self, embedding_layer: tf.keras.Model, *args, **kwargs):
+        """
+        Arguments:
+            embedding_layer: A tf.keras.Model instance that given a batch of
+                images computes their embeddings. It should accept 4D tensors
+                with shape `(batch_size, height, width, num_channels)` as input
+                and output 2D tensors of shape `(batch_size, embedding_size)`.
+        """
+        super().__init__(*args, **kwargs)
+        self.embedding_layer = embedding_layer
```

Comment:
is embedding layer hier niet een verwarrende naam? het is een heel model dat als laatste layer een embedding layer heeft?

Reply:
Ja het ligt er een beetje aan hoe je er tegen aan kijkt. In Keras terminologie is `Model` een subclass van `Layer` en daarmee elk model ook een layer, maar ik snap je punt. Ik kan het hernoemen naar `embedding_model`.